### PR TITLE
Fixed #29108 -- Fixed crash in aggregation of sliced/ordered/distinct querysets.

### PR DESCRIPTION
**Pull request by charettes**
Regression in 4acae21846f6212aa992763e587c7e201828d7b0.
https://code.djangoproject.com/ticket/29108

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -547,7 +547,9 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
                 # to exclude extraneous selects.
                 sub_selects = []
                 sub_params = []
-                for select, _, alias in self.select:
+                for index, (select, _, alias) in enumerate(self.select, start=1):
+                    if not alias and with_col_aliases:
+                        alias = 'Col%d' % index
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: Converting `sub_params + params` to a `tuple` uncovered this issue with the mishandling of `with_col_aliases`.

**Comment by charettes** [comment]
I should be able to look at the PostgreSQL failure later this week.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by charettes** [comment]
@felixxm any idea why Oracle is failing with `inconsistent datatypes: expected - got NCLOB` on the added test?

**Comment by charettes** [comment]
Ahh I think it's because we're trying to do a `DISTINCT` on a text/clob field.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@charettes Exactly 😄

File: docs/releases/2.0.3.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+==========================
+
+*Expected March 1, 2018*
+
+Django 2.0.3 fixes several bugs in 2.0.2.
+
+Bugfixes
+========
+
+* Fixed regression in the use of sliced ``QuerySet.order_by().distinct()``
```
Comment (timgraham) [inline docs/releases/2.0.3.txt]: Fixed regression that caused sliced ``QuerySet.order_by().distinct()`` followed by ``count()`` to crash

File: tests/queries/tests.py [hunk 3]
```diff
@@ -2113,6 +2113,9 @@ def test_distinct_ordered_sliced_subquery(self):
             ).order_by('id').values_list('id', flat=True), [2, 4]
         )
 
+    def test_distinct_ordered_sliced_subquery_aggregation(self):
+        Tag.objects.distinct().order_by('category__name')[:20].count()
```
Comment (timgraham) [inline tests/queries/tests.py]: Add an assertion about the count to verify SQL correctness?

**Comment by felixxm** [comment]
@timgraham IMO this database feature name is misleading because it's false only on MySQL (after 0899d58), which doesn't support sliced subqueries only with `IN/ALL/ANY/SOME` (see [subquery-errors](https://dev.mysql.com/doc/refman/5.7/en/subquery-errors.html)). I prepared PR #9669 to change it.

**Comment by charettes** [comment]
Thanks for dealing with the extra changes required @timgraham, @felixxm. I appreciate it!

**Pull request merged**